clients/src/main/java/org/apache/kafka/common/record/CompressionType.java

```diff
@@ -179,4 +185,17 @@ private static MethodHandle findConstructor(String className, MethodType methodT
         }
     }
 
+    static class RecyclingBufferSupplier implements KafkaLZ4BlockInputStream.BufferSupplier {
+
+        ByteBuffer theBuffer;
+
+        @Override
+        public ByteBuffer get(int size) {
+            if (theBuffer == null || theBuffer.capacity() < size) {
+                theBuffer = ByteBuffer.allocate(size);
+            }
+            theBuffer.limit(size);
+            return theBuffer;
```

Comment:
That's one issue. The other is that if there are two `KafkaLZ4BlockInputStream` instances in the same thread: https://github.com/apache/kafka/pull/3164/commits/8403cbb616dd88cd7b06dfe8910aa0d0cded8b6f#diff-ee888ddca9dfb15863eeddcd379f65c6R36